File: test_lexer_debug.py
#!/usr/bin/env python3
"""
Debug test to see what tokens the lexer produces for function declarations
"""

import sys
import os
sys.path.append(os.path.join(os.path.dirname(__file__), 'runa', 'src'))

from runa.compiler.lexer import RunaLexer

def test_function_declaration_tokens():
    """Test what tokens are produced for function declarations."""
    
    # Test the function declaration that's failing
    source = '''Process called "Calculate Area" that takes width and height returns Integer:
    Set area to width multiplied by height
    Return area'''
    
    print("Testing function declaration tokenization:")
    print(f"Source: {repr(source)}")
    print()
    
    lexer = RunaLexer()
    try:
        tokens = lexer.tokenize(source)
        print("Tokens produced:")
        for i, token in enumerate(tokens):
            print(f"  {i:2d}: {token}")
    except Exception as e:
        print(f"Error: {e}")
    
    print("\n" + "="*50)
    
    # Test a simpler function declaration
    source2 = 'Process called "test" that takes nothing:'
    print(f"Testing simpler function: {repr(source2)}")
    print()
    
    try:
        tokens2 = lexer.tokenize(source2)
        print("Tokens produced:")
        for i, token in enumerate(tokens2):
            print(f"  {i:2d}: {token}")
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    test_function_declaration_tokens() 

File: bootstrap/v0.0.7.6/test_lexer_debug.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Minimal test to see what tokens the lexer produces for "Type called"
extern void* lexer_create(const char* source);
extern void* lexer_next_token(void* lexer);
extern int memory_get_int32(void* ptr, int offset);
extern void* memory_get_pointer(void* ptr, int offset);

int main() {
    const char* source = "Type called \"Point\":\n    x as Integer\nEnd Type\n";

    void* lexer = lexer_create(source);
    printf("Lexer created\n");

    for (int i = 0; i < 15; i++) {
        void* token = lexer_next_token(lexer);
        if (!token) break;

        int token_type = memory_get_int32(token, 0);
        void* value = memory_get_pointer(token, 8);

        printf("Token %d: type=%d", i, token_type);
        if (value) {
            printf(" value='%s'", (char*)value);
        }
        printf("\n");

        if (token_type == 0) break; // EOF
    }

    return 0;
}
